File: packages/mesh-runtime/test/common/host.mjs
// The test host (D10), in Node: the Rust tests' host
// (`crates/mesh-runtime/tests/common/host.rs`). It supplies snapshots,
// keeps every render, relays each event to dispatch with the render it
// names, and records what dispatch gives.
import { dispatch, render } from "../../dist/index.js";
import { canonical } from "./json.mjs";

export class Host {
  constructor({ model, root, templates }) {
    this.model = model;
    this.program = { root, templates };
    /** Every render, in order: the host keeps them all. */
    this.renders = [];
  }

  /** Renders a snapshot, keeping the render. Throws with the diagnostics if it fails. */
  async render(snapshot) {
    const result = await render({ program: this.program, model: this.model, snapshot });
    if (result.diagnostics) {
      throw new Error(JSON.stringify(result.diagnostics));
    }
    this.renders.push(result.render);
    return result.render;
  }

  /**
   * Dispatches `event` of the `occurrence`th node binding it (document
   * order) in render `index`, with that render. The result as the
   * document a host would record: the intent, or the diagnostics.
   */
  async dispatch(index, event, occurrence, payload) {
    const kept = this.renders[index];
    const handlers = [];
    collect(kept.tree.root, event, handlers);
    const result = await dispatch(kept, handlers[occurrence], payload);
    return canonical(result.intent ?? result.diagnostics);
  }
}

function collect(node, event, out) {
  if (event in node.events) out.push(node.events[event]);
  for (const child of node.children) {
    if (child.type === "node") collect(child, event, out);
  }
}

File: crates/mesh-runtime/tests/common/host.rs
//! The test host (D10): supplies snapshots from JSON fixtures, keeps
//! every render, relays each event to dispatch with the render it names,
//! and records what dispatch gives. It models the host's obligation
//! (docs/manual/runtime.md): an event is dispatched with the render whose
//! tree the renderer had drawn.

use mesh_runtime::{HostRecord, HostValue, Program, Render, RuntimeDiagnostic};

pub struct Host {
    pub model: String,
    pub root: String,
    pub templates: Vec<String>,
    /// Every render, in order: the host keeps them all.
    pub renders: Vec<Render>,
}

impl Host {
    /// Renders from a snapshot fixture, keeping the render.
    pub fn render(&mut self, snapshot: &str) -> Result<&Render, Vec<RuntimeDiagnostic>> {
        let record = HostRecord::from_json(snapshot).expect("the fixture is JSON");
        let texts: Vec<&str> = self.templates.iter().map(String::as_str).collect();
        let render = mesh_runtime::render(
            &Program {
                root: &self.root,
                templates: &texts,
            },
            &self.model,
            &record,
        )?;
        self.renders.push(render);
        Ok(self.renders.last().unwrap())
    }

    /// Dispatches `event` of the `occurrence`th node binding it (document
    /// order) in render `render`, with that render, and `payload`. The
    /// result as the document a host would record: the intent, or the
    /// runtime diagnostics document.
    pub fn dispatch(
        &self,
        render: usize,
        event: &str,
        occurrence: usize,
        payload: Option<&str>,
    ) -> String {
        let kept = &self.renders[render];
        let mut handlers = Vec::new();
        collect(&kept.tree().root, event, &mut handlers);
        let handler = &handlers[occurrence];
        let payload = payload.map(|text| HostValue::from_json(text).expect("the payload is JSON"));
        match mesh_runtime::dispatch(kept, handler, payload.as_ref()) {
            Ok(intent) => intent.to_json(),
            Err(diagnostics) => mesh_runtime::to_json(&diagnostics, &self.model),
        }
    }
}

fn collect(node: &mesh_runtime::Node, event: &str, out: &mut Vec<String>) {
    if let Some(handler) = node.events.get(event) {
        out.push(handler.clone());
    }
    for child in &node.children {
        if let mesh_runtime::TreeChild::Node(child) = child {
            collect(child, event, out);
        }
    }
}
